bmi_sdk/bmi_sdk/bmi_variable.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+from enum import Enum
+
+# Once python 3.8 is no longer supported, just get Annotated directly
+# from typing import Annotated, Any, Literal, Optional, Self, Union
+from typing import Any, Literal, Optional, Union
+
+import numpy as np
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    RootModel,
+    computed_field,
+    model_validator,
+)
+
+# Only needed on python 3.8
+from typing_extensions import Annotated, Self
+
+from .bmi_grid import Grid
+
+Integer = RootModel[Annotated[str, Literal["int", "int16", "int32", "integer"]]]
+Long = RootModel[Annotated[str, Literal["int", "int64", "longlong"]]]
+Float = RootModel[
+    Annotated[
+        str,
+        Literal[
+            "float",
+            "float32",
+            "np.float32",
+            "numpy.float32",
+            "np.single",
+            "numpy.single",
+        ],
+    ]
+]
+Double = RootModel[
+    Annotated[str, Literal["float", "float64", "np.float64", "numpy.float64"]]
+]
+
+VarType = Union[Integer, Long, Float, Double]
+
+
+class VarLocation(str, Enum):
+    """Enumeration of variable grid location strings supported by BMI.
+
+    See https://bmi.readthedocs.io/en/stable/#get-var-location
+
+    """
+
+    node = "node"
+    edge = "edge"
+    face = "face"
+
+
+class BmiVariable(BaseModel):
+    """Abstraction of a BMI variable which contains all the standard BMI variable metadata."""
+
+    # Required for Grid attribute
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    name: str = Field(min_length=1)
+    grid: Annotated[Grid, "Bmi Grid object"] = Field(default=Grid(0, 0, "scalar"))
+    type: VarType = Field(default="float")
+    units: str = Field(default="-")
+    # By default, value is a numpy scalar of float type
+    value: Optional[Any] = Field(default=np.zeros(tuple()), dtype=np.float32)
```

Comment:
Same here with needing to use `default_factory`. We should constraint this field type more. I need to look at the PEP for variadic generic type parameters again, but we should be able to use that here to say it's some sequence of shape R, C.

Reply:
I definitely punted on this one a bit for the time being.  I almost started writing a `BmiNDvar` sublcass, then thought better of that and just went back this for the time being.  If you find some corresponding documentation related, would love to see it!